File: lib/evaluator.ts
// Compact, dependency-free 5-to-7 card poker hand evaluator.
//
// Returns a single integer score where a HIGHER score is a better hand. Scores
// are comparable across all hands, so equity can be computed by counting wins,
// ties and losses. The scheme packs the hand category and kickers into one int:
//
//   score = (category << 20) | tiebreakers
//
// category: 8 = straight flush ... 0 = high card.
// This is a classic rank-histogram evaluator: correct and fast enough to drive
// Monte Carlo in a Web Worker (millions of evals/sec), with no lookup-table
// asset to ship. It can later be swapped for a WASM evaluator behind the same
// signature without touching callers.

import { Card, cardRank, cardSuit } from './cards';

export const HAND_CATEGORY_NAMES = [
  'High Card',
  'Pair',
  'Two Pair',
  'Three of a Kind',
  'Straight',
  'Flush',
  'Full House',
  'Four of a Kind',
  'Straight Flush',
] as const;

const STRAIGHT_FLUSH = 8;
const FOUR_KIND = 7;
const FULL_HOUSE = 6;
const FLUSH = 5;
const STRAIGHT = 4;
const THREE_KIND = 3;
const TWO_PAIR = 2;
const ONE_PAIR = 1;
const HIGH_CARD = 0;

/**
 * Evaluate the best 5-card hand from 5, 6 or 7 cards.
 * @param cards array of card ints (0..51)
 * @returns integer score (higher is better)
 */
export function evaluate(cards: Card[]): number {
  // Rank counts and per-suit rank bitmasks.
  const rankCount = new Int8Array(13);
  const suitRankMask = [0, 0, 0, 0]; // bitmask of ranks present per suit
  let rankMask = 0;

  for (let i = 0; i < cards.length; i++) {
    const c = cards[i];
    const r = cardRank(c);
    const s = cardSuit(c);
    rankCount[r]++;
    suitRankMask[s] |= 1 << r;
    rankMask |= 1 << r;
  }

  // Flush / straight flush detection
  for (let s = 0; s < 4; s++) {
    const mask = suitRankMask[s];
    if (popcount(mask) >= 5) {
      const sf = straightHighFromMask(mask);
      if (sf >= 0) return score(STRAIGHT_FLUSH, sf);
      // Plain flush: top 5 ranks in this suit
      return score(FLUSH, topN(mask, 5));
    }
  }

  // Group ranks by count (quads, trips, pairs)
  let quad = -1;
  const trips: number[] = [];
  const pairs: number[] = [];
  for (let r = 12; r >= 0; r--) {
    const n = rankCount[r];
    if (n === 4) quad = r;
    else if (n === 3) trips.push(r);
    else if (n === 2) pairs.push(r);
  }

  if (quad >= 0) {
    const kicker = highestExcept(rankMask, quad);
    return score(FOUR_KIND, (quad << 4) | kicker);
  }

  if (trips.length >= 1 && (trips.length >= 2 || pairs.length >= 1)) {
    const three = trips[0];
    const pair = trips.length >= 2 ? trips[1] : pairs[0];
    return score(FULL_HOUSE, (three << 4) | pair);
  }

  // Straight (non-flush)
  const st = straightHighFromMask(rankMask);
  if (st >= 0) return score(STRAIGHT, st);

  if (trips.length >= 1) {
    const three = trips[0];
    return score(THREE_KIND, (three << 16) | topNExcept(rankMask, 2, three));
  }

  if (pairs.length >= 2) {
    const hi = pairs[0];
    const lo = pairs[1];
    const kicker = highestExcept(rankMask, hi, lo);
    return score(TWO_PAIR, (hi << 8) | (lo << 4) | kicker);
  }

  if (pairs.length === 1) {
    const p = pairs[0];
    return score(ONE_PAIR, (p << 16) | topNExcept(rankMask, 3, p));
  }

  return score(HIGH_CARD, topN(rankMask, 5));
}

function score(category: number, tiebreak: number): number {
  return (category << 24) | tiebreak;
}

export function handCategory(score: number): number {
  return score >>> 24;
}

// --- bit helpers ---------------------------------------------------------

function popcount(x: number): number {
  let c = 0;
  while (x) {
    x &= x - 1;
    c++;
  }
  return c;
}

/**
 * Given a rank bitmask, return the high card rank of the best straight, or -1.
 * Handles the wheel (A-2-3-4-5) where the ace plays low.
 */
function straightHighFromMask(mask: number): number {
  // Scan from ace-high (T-J-Q-K-A) down to 6-high; return the first hit.
  for (let hi = 12; hi >= 4; hi--) {
    const need = 0b11111 << (hi - 4);
    if ((mask & need) === need) return hi;
  }
  // Wheel: A-2-3-4-5 (bits 12,0,1,2,3), ace plays low -> 5-high (rank index 3).
  const wheel = (1 << 12) | (1 << 0) | (1 << 1) | (1 << 2) | (1 << 3);
  if ((mask & wheel) === wheel) return 3;
  return -1;
}

/** Top N ranks from a bitmask packed 4 bits each (highest first). */
function topN(mask: number, n: number): number {
  let out = 0;
  let count = 0;
  for (let r = 12; r >= 0 && count < n; r--) {
    if (mask & (1 << r)) {
      out = (out << 4) | r;
      count++;
    }
  }
  return out;
}

/** Top N ranks excluding a specific rank. */
function topNExcept(mask: number, n: number, except: number): number {
  return topN(mask & ~(1 << except), n);
}

function highestExcept(mask: number, ...except: number[]): number {
  let m = mask;
  for (const e of except) m &= ~(1 << e);
  for (let r = 12; r >= 0; r--) if (m & (1 << r)) return r;
  return 0;
}

File: preflop-solver/src/evaluator.rs
//! Compact 5-to-7 card evaluator adapted from `wasm/src/eval.rs`.
//! Higher scores represent stronger hands.

#[inline]
fn card_rank(card: u8) -> u32 {
    (card >> 2) as u32
}

#[inline]
fn card_suit(card: u8) -> usize {
    (card & 3) as usize
}

const STRAIGHT_FLUSH: u32 = 8;
const FOUR_KIND: u32 = 7;
const FULL_HOUSE: u32 = 6;
const FLUSH: u32 = 5;
const STRAIGHT: u32 = 4;
const THREE_KIND: u32 = 3;
const TWO_PAIR: u32 = 2;
const ONE_PAIR: u32 = 1;
const HIGH_CARD: u32 = 0;

#[inline]
fn score(category: u32, tiebreak: u32) -> u32 {
    (category << 24) | tiebreak
}

pub fn evaluate(cards: &[u8]) -> u32 {
    assert!((5..=7).contains(&cards.len()));
    let mut rank_count = [0u8; 13];
    let mut suit_rank_mask = [0u32; 4];
    let mut rank_mask = 0u32;

    for &card in cards {
        let rank = card_rank(card);
        let suit = card_suit(card);
        rank_count[rank as usize] += 1;
        suit_rank_mask[suit] |= 1 << rank;
        rank_mask |= 1 << rank;
    }

    for mask in suit_rank_mask {
        if mask.count_ones() >= 5 {
            let straight_flush = straight_high(mask);
            if straight_flush >= 0 {
                return score(STRAIGHT_FLUSH, straight_flush as u32);
            }
            return score(FLUSH, top_n(mask, 5));
        }
    }

    let mut quad = -1i32;
    // Seven cards can contain at most two trip ranks or three pair ranks.
    // Keep these tiny collections on the stack: this evaluator is called
    // millions of times while building postflop information-set buckets.
    let mut trips = [0u32; 2];
    let mut trip_count = 0usize;
    let mut pairs = [0u32; 3];
    let mut pair_count = 0usize;
    for rank in (0..13u32).rev() {
        match rank_count[rank as usize] {
            4 => quad = rank as i32,
            3 => {
                trips[trip_count] = rank;
                trip_count += 1;
            }
            2 => {
                pairs[pair_count] = rank;
                pair_count += 1;
            }
            _ => {}
        }
    }

    if quad >= 0 {
        let kicker = highest_except(rank_mask, &[quad as u32]);
        return score(FOUR_KIND, ((quad as u32) << 4) | kicker);
    }
    if trip_count > 0 && (trip_count >= 2 || pair_count > 0) {
        let pair = if trip_count >= 2 { trips[1] } else { pairs[0] };
        return score(FULL_HOUSE, (trips[0] << 4) | pair);
    }
    let straight = straight_high(rank_mask);
    if straight >= 0 {
        return score(STRAIGHT, straight as u32);
    }
    if trip_count > 0 {
        return score(
            THREE_KIND,
            (trips[0] << 16) | top_n_except(rank_mask, 2, trips[0]),
        );
    }
    if pair_count >= 2 {
        let kicker = highest_except(rank_mask, &[pairs[0], pairs[1]]);
        return score(TWO_PAIR, (pairs[0] << 8) | (pairs[1] << 4) | kicker);
    }
    if pair_count == 1 {
        return score(
            ONE_PAIR,
            (pairs[0] << 16) | top_n_except(rank_mask, 3, pairs[0]),
        );
    }
    score(HIGH_CARD, top_n(rank_mask, 5))
}

fn straight_high(mask: u32) -> i32 {
    for high in (4..=12i32).rev() {
        let needed = 0b11111u32 << (high - 4);
        if (mask & needed) == needed {
            return high;
        }
    }
    let wheel = (1 << 12) | (1 << 0) | (1 << 1) | (1 << 2) | (1 << 3);
    if (mask & wheel) == wheel {
        return 3;
    }
    -1
}

fn top_n(mask: u32, count: usize) -> u32 {
    let mut result = 0u32;
    let mut found = 0;
    for rank in (0..13u32).rev() {
        if mask & (1 << rank) != 0 {
            result = (result << 4) | rank;
            found += 1;
            if found == count {
                break;
            }
        }
    }
    result
}

fn top_n_except(mask: u32, count: usize, except: u32) -> u32 {
    top_n(mask & !(1 << except), count)
}

fn highest_except(mask: u32, except: &[u32]) -> u32 {
    let mut available = mask;
    for &rank in except {
        available &= !(1 << rank);
    }
    for rank in (0..13u32).rev() {
        if available & (1 << rank) != 0 {
            return rank;
        }
    }
    0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ranks_known_seven_card_hands() {
        let straight_flush = evaluate(&[51, 47, 43, 39, 35, 0, 1]);
        let quads = evaluate(&[48, 49, 50, 51, 47, 0, 1]);
        let full_house = evaluate(&[48, 49, 50, 44, 45, 0, 1]);
        assert!(straight_flush > quads);
        assert!(quads > full_house);
    }

    #[test]
    fn wheel_is_lower_than_six_high_straight() {
        let wheel = evaluate(&[48, 0, 5, 10, 15]);
        let six_high = evaluate(&[0, 5, 10, 15, 16]);
        assert!(six_high > wheel);
    }

    #[test]
    fn seven_card_rank_buffers_cover_two_trips_and_three_pairs() {
        let two_trips = evaluate(&[48, 49, 50, 44, 45, 46, 40]);
        assert_eq!(two_trips, score(FULL_HOUSE, (12 << 4) | 11));

        let three_pairs = evaluate(&[48, 49, 44, 45, 40, 41, 36]);
        assert_eq!(three_pairs, score(TWO_PAIR, (12 << 8) | (11 << 4) | 10));
    }
}
